File: web_dashboard/pages/ticker_details.py
#!/usr/bin/env python3
"""
Ticker Details Page
===================

Comprehensive ticker information page that aggregates data from all databases
and provides external links to financial websites.
"""

import streamlit as st
import sys
from pathlib import Path
from typing import Dict, Optional, Any
from datetime import datetime, timezone
import pandas as pd
import logging

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))

from auth_utils import is_authenticated, get_user_email
from navigation import render_navigation
from postgres_client import PostgresClient
from supabase_client import SupabaseClient
from ticker_utils import get_ticker_info, get_ticker_external_links

logger = logging.getLogger(__name__)

# Page configuration
st.set_page_config(
    page_title="Ticker Details",
    page_icon="📊",
    layout="wide"
)

# Check authentication
if not is_authenticated():
    st.switch_page("streamlit_app.py")
    st.stop()

# Refresh token if needed
from auth_utils import refresh_token_if_needed
if not refresh_token_if_needed():
    from auth_utils import logout_user
    logout_user()
    st.error("Your session has expired. Please log in again.")
    st.switch_page("streamlit_app.py")
    st.stop()

# Render navigation
render_navigation(show_ai_assistant=True, show_settings=True)

# Initialize clients
@st.cache_resource
def get_postgres_client():
    """Get Postgres client instance"""
    try:
        return PostgresClient()
    except Exception as e:
        logger.error(f"Failed to initialize PostgresClient: {e}")
        return None

@st.cache_resource
def get_supabase_client():
    """Get Supabase client instance"""
    try:
        return SupabaseClient(use_service_role=True)
    except Exception as e:
        logger.error(f"Failed to initialize SupabaseClient: {e}")
        return None

postgres_client = get_postgres_client()
supabase_client = get_supabase_client()

# Get ticker from query parameters
query_params = st.query_params
ticker = query_params.get("ticker", "")

# Ticker search box
col_search, col_spacer = st.columns([0.3, 0.7])
with col_search:
    search_ticker = st.text_input(
        "Search Ticker",
        value=ticker.upper() if ticker else "",
        placeholder="Enter ticker symbol (e.g., AAPL)",
        key="ticker_search"
    ).upper().strip()

# If user entered a new ticker, update query params
if search_ticker and search_ticker != ticker:
    st.query_params["ticker"] = search_ticker
    st.rerun()

# Use the ticker from query params or search
current_ticker = ticker.upper().strip() if ticker else search_ticker

if not current_ticker:
    st.title("📊 Ticker Details")
    st.info("Enter a ticker symbol above to view detailed information.")
    st.stop()

# Fetch ticker information
@st.cache_data(ttl=60)
def fetch_ticker_data(_ticker: str):
    """Fetch ticker data with caching"""
    # Get clients inside the cached function
    pg_client = get_postgres_client()
    sb_client = get_supabase_client()
    return get_ticker_info(_ticker, sb_client, pg_client)

# Check if clients are available
if not postgres_client and not supabase_client:
    st.error("⚠️ Unable to connect to databases. Please check your configuration.")
    st.stop()

# Helper function to format dates safely
def format_date_safe(date_val):
    """Safely format a date value that might be string or datetime"""
    if not date_val:
        return 'N/A'
    if isinstance(date_val, str):
        return date_val[:10]  # Return first 10 chars (YYYY-MM-DD)
    try:
        return date_val.strftime('%Y-%m-%d')  # Format datetime object
    except (AttributeError, ValueError):
        return str(date_val)[:10]

try:
    with st.spinner(f"Loading information for {current_ticker}..."):
        ticker_data = fetch_ticker_data(current_ticker)
except Exception as e:
    logger.error(f"Error fetching ticker data for {current_ticker}: {e}", exc_info=True)
    st.error(f"❌ Error loading ticker data: {str(e)}")
    st.info("Please try again or contact support if the problem persists.")
    st.stop()

# Header
st.title(f"📊 {current_ticker}")

# Wrap main content in try-except for better error handling
try:
    # Basic Info Section
    basic_info = ticker_data.get('basic_info')
if basic_info:
    company_name = basic_info.get('company_name', 'N/A')
    sector = basic_info.get('sector', 'N/A')
    industry = basic_info.get('industry', 'N/A')
    currency = basic_info.get('currency', 'USD')
    exchange = basic_info.get('exchange', 'N/A')
    
    st.header(f"{company_name}")
    col1, col2, col3 = st.columns(3)
    with col1:
        st.metric("Sector", sector)
    with col2:
        st.metric("Industry", industry)
    with col3:
        st.metric("Currency", currency)
    
    if exchange != 'N/A':
        st.caption(f"Exchange: {exchange}")
else:
    st.info(f"Basic information not found for {current_ticker} in database.")

# External Links Section
st.header("🔗 External Links")
external_links = get_ticker_external_links(
    current_ticker,
    exchange=basic_info.get('exchange') if basic_info else None
)

# Display links in columns
cols = st.columns(4)
link_items = list(external_links.items())
for i, (name, url) in enumerate(link_items):
    with cols[i % 4]:
        st.markdown(f"[{name}]({url})")

st.markdown("---")

# Portfolio Data Section
portfolio_data = ticker_data.get('portfolio_data')
if portfolio_data and (portfolio_data.get('has_positions') or portfolio_data.get('has_trades')):
    st.header("💼 Portfolio Data")
    
    # Current Positions
    if portfolio_data.get('has_positions'):
        positions = portfolio_data.get('positions', [])
        if positions:
            st.subheader("Current Positions")
            # Get latest position for each fund
            latest_positions = {}
            for pos in positions:
                fund = pos.get('fund', 'Unknown')
                if fund not in latest_positions:
                    latest_positions[fund] = pos
                else:
                    # Keep the most recent
                    if pos.get('date', '') > latest_positions[fund].get('date', ''):
                        latest_positions[fund] = pos
            
            pos_df = pd.DataFrame([
                {
                    'Fund': pos.get('fund', 'N/A'),
                    'Shares': f"{pos.get('shares', 0):,.2f}",
                    'Price': f"${pos.get('price', 0):.2f}",
                    'Cost Basis': f"${pos.get('cost_basis', 0):.2f}",
                    'P&L': f"${pos.get('pnl', 0):.2f}",
                    'Date': pos.get('date', 'N/A')[:10] if pos.get('date') else 'N/A'
                }
                for pos in latest_positions.values()
            ])
            st.dataframe(pos_df, use_container_width=True, hide_index=True)
    
    # Trade History
    if portfolio_data.get('has_trades'):
        trades = portfolio_data.get('trades', [])
        if trades:
            st.subheader("Recent Trade History")
            trade_df = pd.DataFrame([
                {
                    'Date': trade.get('date', 'N/A')[:10] if trade.get('date') else 'N/A',
                    'Action': trade.get('action', 'N/A'),
                    'Shares': f"{trade.get('shares', 0):,.2f}",
                    'Price': f"${trade.get('price', 0):.2f}",
                    'Fund': trade.get('fund', 'N/A'),
                    'Reason': trade.get('reason', 'N/A')[:50] if trade.get('reason') else 'N/A'
                }
                for trade in trades[:20]  # Show last 20 trades
            ])
            st.dataframe(trade_df, use_container_width=True, hide_index=True)
else:
    st.info(f"No portfolio data found for {current_ticker}.")

st.markdown("---")

# Research Articles Section
research_articles = ticker_data.get('research_articles', [])
if research_articles:
    st.header("📚 Research Articles")
    st.caption(f"Found {len(research_articles)} articles mentioning {current_ticker} (last 30 days)")
    
    for article in research_articles[:10]:  # Show top 10
        with st.expander(f"{article.get('title', 'Untitled')[:80]}..."):
            col1, col2 = st.columns([3, 1])
            with col1:
                if article.get('summary'):
                    st.write(article.get('summary', '')[:500] + '...' if len(article.get('summary', '')) > 500 else article.get('summary', ''))
                if article.get('url'):
                    st.markdown(f"[Read Full Article]({article.get('url')})")
            with col2:
                st.caption(f"Source: {article.get('source', 'Unknown')}")
                if article.get('published_at'):
                    published_date = format_date_safe(article.get('published_at'))
                    st.caption(f"Published: {published_date}")
                if article.get('sentiment'):
                    st.caption(f"Sentiment: {article.get('sentiment', 'N/A')}")
else:
    st.info(f"No research articles found for {current_ticker} (last 30 days).")

st.markdown("---")

# Social Sentiment Section
social_sentiment = ticker_data.get('social_sentiment')
if social_sentiment:
    st.header("💬 Social Sentiment")
    
    latest_metrics = social_sentiment.get('latest_metrics', [])
    if latest_metrics:
        st.subheader("Latest Metrics")
        metrics_df = pd.DataFrame([
            {
                'Platform': metric.get('platform', 'N/A').title(),
                'Sentiment': metric.get('sentiment_label', 'N/A'),
                'Score': f"{metric.get('sentiment_score', 0):.2f}",
                'Volume': metric.get('volume', 0),
                'Bull/Bear Ratio': f"{metric.get('bull_bear_ratio', 0):.2f}" if metric.get('bull_bear_ratio') else 'N/A',
                'Last Updated': format_date_safe(metric.get('created_at')) if metric.get('created_at') else 'N/A'
            }
            for metric in latest_metrics
        ])
        st.dataframe(metrics_df, use_container_width=True, hide_index=True)
    
    alerts = social_sentiment.get('alerts', [])
    if alerts:
        st.subheader("Recent Alerts (Last 24 Hours)")
        for alert in alerts:
            sentiment_label = alert.get('sentiment_label', 'N/A')
            if sentiment_label == 'EUPHORIC':
                st.success(f"**{alert.get('platform', 'Unknown').title()}** - {sentiment_label} (Score: {alert.get('sentiment_score', 0):.2f})")
            elif sentiment_label == 'FEARFUL':
                st.error(f"**{alert.get('platform', 'Unknown').title()}** - {sentiment_label} (Score: {alert.get('sentiment_score', 0):.2f})")
            elif sentiment_label == 'BULLISH':
                st.info(f"**{alert.get('platform', 'Unknown').title()}** - {sentiment_label} (Score: {alert.get('sentiment_score', 0):.2f})")
else:
    st.info(f"No social sentiment data available for {current_ticker}.")

st.markdown("---")

# Congress Trades Section
congress_trades = ticker_data.get('congress_trades', [])
if congress_trades:
    st.header("🏛️ Congress Trades")
    st.caption(f"Found {len(congress_trades)} recent trades by politicians (last 30 days)")
    
    trades_df = pd.DataFrame([
        {
            'Date': trade.get('transaction_date', 'N/A'),
            'Politician': trade.get('politician', 'N/A'),
            'Chamber': trade.get('chamber', 'N/A'),
            'Type': trade.get('type', 'N/A'),
            'Amount': trade.get('amount', 'N/A'),
            'Party': trade.get('party', 'N/A')
        }
        for trade in congress_trades[:20]  # Show last 20
    ])
    st.dataframe(trades_df, use_container_width=True, hide_index=True)
else:
    st.info(f"No congress trades found for {current_ticker} (last 30 days).")

st.markdown("---")

# Watchlist Status Section
watchlist_status = ticker_data.get('watchlist_status')
if watchlist_status:
    st.header("📋 Watchlist Status")
    col1, col2, col3 = st.columns(3)
    with col1:
        st.metric("Status", "✅ In Watchlist" if watchlist_status.get('is_active') else "❌ Not Active")
    with col2:
        st.metric("Priority Tier", watchlist_status.get('priority_tier', 'N/A'))
    with col3:
        st.metric("Source", watchlist_status.get('source', 'N/A'))
else:
    st.info(f"{current_ticker} is not in the watchlist.")

except Exception as e:
    logger.error(f"Error rendering ticker details page for {current_ticker}: {e}", exc_info=True)
    st.error(f"❌ An error occurred while displaying ticker information: {str(e)}")
    st.info("Please try refreshing the page or contact support if the problem persists.")

# Footer
st.markdown("---")
st.caption(f"Data last updated: {datetime.now(timezone.utc).strftime('%Y-%m-%d %H:%M:%S UTC')}")

File: web_dashboard/static/js/ticker_details.js
// Ticker Details Page JavaScript

let currentTicker = '';
let tickerList = [];

// Initialize page on load
document.addEventListener('DOMContentLoaded', function() {
    // Get ticker from URL query parameter
    const urlParams = new URLSearchParams(window.location.search);
    const tickerParam = urlParams.get('ticker');
    
    // Load ticker list first
    loadTickerList().then(() => {
        // If ticker in URL, load it
        if (tickerParam) {
            currentTicker = tickerParam.toUpperCase();
            document.getElementById('ticker-select').value = currentTicker;
            loadTickerData(currentTicker);
        } else {
            // Show placeholder
            showPlaceholder();
        }
    });
    
    // Set up ticker dropdown change handler
    document.getElementById('ticker-select').addEventListener('change', handleTickerSearch);
    
    // Set up chart controls
    document.getElementById('solid-lines-checkbox').addEventListener('change', function() {
        if (currentTicker) {
            loadAndRenderChart(currentTicker, this.checked);
        }
    });
});

// Load ticker list for dropdown
async function loadTickerList() {
    try {
        const response = await fetch('/api/v2/ticker/list', {
            credentials: 'include'
        });
        
        if (!response.ok) {
            throw new Error('Failed to load ticker list');
        }
        
        const data = await response.json();
        tickerList = data.tickers || [];
        
        // Populate dropdown
        const select = document.getElementById('ticker-select');
        // Clear existing options except first one
        while (select.options.length > 1) {
            select.remove(1);
        }
        
        // Add tickers
        tickerList.forEach(ticker => {
            const option = document.createElement('option');
            option.value = ticker;
            option.textContent = ticker;
            select.appendChild(option);
        });
    } catch (error) {
        console.error('Error loading ticker list:', error);
    }
}

// Handle ticker search dropdown change
function handleTickerSearch() {
    const selectedTicker = document.getElementById('ticker-select').value.toUpperCase().trim();
    
    if (selectedTicker) {
        // Update URL without reload
        const url = new URL(window.location);
        url.searchParams.set('ticker', selectedTicker);
        window.history.pushState({}, '', url);
        
        currentTicker = selectedTicker;
        loadTickerData(selectedTicker);
    } else {
        // Clear URL and show placeholder
        const url = new URL(window.location);
        url.searchParams.delete('ticker');
        window.history.pushState({}, '', url);
        showPlaceholder();
    }
}

// Load all ticker data
async function loadTickerData(ticker) {
    hideAllSections();
    showLoading();
    hideError();
    hidePlaceholder();
    
    try {
        const response = await fetch(`/api/v2/ticker/info?ticker=${encodeURIComponent(ticker)}`, {
            credentials: 'include'
        });
        
        if (!response.ok) {
            const errorData = await response.json();
            throw new Error(errorData.error || 'Failed to load ticker data');
        }
        
        const data = await response.json();
        
        // Render all sections
        renderBasicInfo(data.basic_info);
        renderExternalLinks(data.basic_info);
        renderPortfolioData(data.portfolio_data);
        renderResearchArticles(data.research_articles || []);
        renderSocialSentiment(data.social_sentiment);
        renderCongressTrades(data.congress_trades || []);
        renderWatchlistStatus(data.watchlist_status);
        
        // Load and render chart
        const useSolid = document.getElementById('solid-lines-checkbox').checked;
        loadAndRenderChart(ticker, useSolid);
        
        hideLoading();
    } catch (error) {
        console.error('Error loading ticker data:', error);
        showError(error.message);
        hideLoading();
    }
}

// Render basic info section
function renderBasicInfo(basicInfo) {
    if (!basicInfo) {
        return;
    }
    
    const section = document.getElementById('basic-info-section');
    section.classList.remove('section-hidden');
    
    document.getElementById('company-name').textContent = basicInfo.company_name || 'N/A';
    document.getElementById('sector').textContent = basicInfo.sector || 'N/A';
    document.getElementById('industry').textContent = basicInfo.industry || 'N/A';
    document.getElementById('currency').textContent = basicInfo.currency || 'USD';
    
    const exchangeInfo = document.getElementById('exchange-info');
    if (basicInfo.exchange && basicInfo.exchange !== 'N/A') {
        exchangeInfo.textContent = `Exchange: ${basicInfo.exchange}`;
        exchangeInfo.style.display = 'block';
    } else {
        exchangeInfo.style.display = 'none';
    }
}

// Render external links
async function renderExternalLinks(basicInfo) {
    if (!basicInfo || !basicInfo.ticker) {
        return;
    }
    
    try {
        const exchange = basicInfo.exchange || null;
        const response = await fetch(`/api/v2/ticker/external-links?ticker=${encodeURIComponent(basicInfo.ticker)}${exchange ? `&exchange=${encodeURIComponent(exchange)}` : ''}`, {
            credentials: 'include'
        });
        
        if (!response.ok) {
            return;
        }
        
        const links = await response.json();
        const grid = document.getElementById('external-links-grid');
        grid.innerHTML = '';
        
        Object.entries(links).forEach(([name, url]) => {
            const link = document.createElement('a');
            link.href = url;
            link.target = '_blank';
            link.rel = 'noopener noreferrer';
            link.className = 'external-link';
            link.textContent = name;
            grid.appendChild(link);
        });
        
        if (Object.keys(links).length > 0) {
            document.getElementById('external-links-section').classList.remove('section-hidden');
        }
    } catch (error) {
        console.error('Error loading external links:', error);
    }
}

// Render portfolio data
function renderPortfolioData(portfolioData) {
    if (!portfolioData || (!portfolioData.has_positions && !portfolioData.has_trades)) {
        return;
    }
    
    const section = document.getElementById('portfolio-section');
    section.classList.remove('section-hidden');
    
    // Render positions
    if (portfolioData.has_positions && portfolioData.positions && portfolioData.positions.length > 0) {
        const tbody = document.getElementById('positions-tbody');
        tbody.innerHTML = '';
        
        // Get latest position per fund
        const latestPositions = {};
        portfolioData.positions.forEach(pos => {
            const fund = pos.fund || 'Unknown';
            if (!latestPositions[fund] || (pos.date && pos.date > (latestPositions[fund].date || ''))) {
                latestPositions[fund] = pos;
            }
        });
        
        Object.values(latestPositions).forEach(pos => {
            const row = document.createElement('tr');
            row.innerHTML = `
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${pos.fund || 'N/A'}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatNumber(pos.shares || 0, 2)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatCurrency(pos.price || 0)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatCurrency(pos.cost_basis || 0)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm ${(pos.pnl || 0) >= 0 ? 'text-green-600' : 'text-red-600'}">${formatCurrency(pos.pnl || 0)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-500">${formatDate(pos.date)}</td>
            `;
            tbody.appendChild(row);
        });
        
        document.getElementById('positions-container').style.display = 'block';
    } else {
        document.getElementById('positions-container').style.display = 'none';
    }
    
    // Render trades
    if (portfolioData.has_trades && portfolioData.trades && portfolioData.trades.length > 0) {
        const tbody = document.getElementById('trades-tbody');
        tbody.innerHTML = '';
        
        portfolioData.trades.slice(0, 20).forEach(trade => {
            const row = document.createElement('tr');
            row.innerHTML = `
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatDate(trade.date)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.action || 'N/A'}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatNumber(trade.shares || 0, 2)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatCurrency(trade.price || 0)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.fund || 'N/A'}</td>
                <td class="px-6 py-4 text-sm text-gray-500">${(trade.reason || 'N/A').substring(0, 50)}</td>
            `;
            tbody.appendChild(row);
        });
    }
}

// Load and render chart
async function loadAndRenderChart(ticker, useSolid) {
    // Show loading indicator
    const chartLoading = document.getElementById('chart-loading');
    const chartContainer = document.getElementById('chart-container');
    
    // Clear any existing chart
    chartContainer.innerHTML = '';
    chartLoading.classList.remove('section-hidden');
    
    // Show chart section (but with loading indicator)
    document.getElementById('chart-section').classList.remove('section-hidden');
    
    try {
        // Detect actual theme from page
        const htmlElement = document.documentElement;
        const dataTheme = htmlElement.getAttribute('data-theme') || 'system';
        let theme = 'light'; // default
        
        if (dataTheme === 'dark') {
            theme = 'dark';
        } else if (dataTheme === 'light') {
            theme = 'light';
        } else if (dataTheme === 'system') {
            // For 'system', check if page is actually in dark mode via CSS
            const bodyBg = window.getComputedStyle(document.body).backgroundColor;
            // Check for dark mode background colors
            const isDark = bodyBg && (
                bodyBg.includes('rgb(31, 41, 55)') ||  // --bg-primary dark
                bodyBg.includes('rgb(17, 24, 39)') ||  // --bg-secondary dark  
                bodyBg.includes('rgb(55, 65, 81)')     // --bg-tertiary dark
            );
            theme = isDark ? 'dark' : 'light';
        }
        
        console.log('Detected theme:', theme, 'from data-theme:', dataTheme);
        
        const response = await fetch(`/api/v2/ticker/chart?ticker=${encodeURIComponent(ticker)}&use_solid=${useSolid}&theme=${encodeURIComponent(theme)}`, {
            credentials: 'include'
        });
        
        // Check if response is JSON
        const contentType = response.headers.get('content-type');
        const isJson = contentType && contentType.includes('application/json');
        
        if (!response.ok) {
            let errorMessage = `Failed to load chart (${response.status})`;
            if (isJson) {
                try {
                    const errorData = await response.json();
                    errorMessage = errorData.error || errorMessage;
                } catch (e) {
                    // If JSON parsing fails, use default message
                }
            } else {
                // Response is HTML (likely an error page)
                errorMessage = `Server error: ${response.status} ${response.statusText}`;
            }
            throw new Error(errorMessage);
        }
        
        if (!isJson) {
            throw new Error('Server returned non-JSON response. Please check your authentication.');
        }
        
        const chartData = await response.json();
        
        // Validate chart data structure
        if (!chartData || !chartData.data || !chartData.layout) {
            throw new Error('Invalid chart data received from server');
        }
        
        // Render with Plotly
        Plotly.newPlot('chart-container', chartData.data, chartData.layout, {responsive: true});
        
        // Hide loading indicator AFTER successful rendering
        chartLoading.classList.add('section-hidden');
        chartLoading.style.display = 'none';
        
        // Load price history for metrics
        loadPriceHistoryMetrics(ticker);
    } catch (error) {
        console.error('Error loading chart:', error);
        // Hide loading indicator
        chartLoading.classList.add('section-hidden');
        // Show error message to user
        showError(`Failed to load chart: ${error.message}`);
        // Hide chart section on error
        document.getElementById('chart-section').classList.add('section-hidden');
    }
}

// Load price history for metrics
async function loadPriceHistoryMetrics(ticker) {
    try {
        const response = await fetch(`/api/v2/ticker/price-history?ticker=${encodeURIComponent(ticker)}&days=90`, {
            credentials: 'include'
        });
        
        if (!response.ok) {
            return;
        }
        
        const data = await response.json();
        const prices = data.data || [];
        
        if (prices.length > 0) {
            const firstPrice = prices[0].price || 0;
            const lastPrice = prices[prices.length - 1].price || 0;
            const priceChange = lastPrice - firstPrice;
            const priceChangePct = firstPrice > 0 ? (priceChange / firstPrice * 100) : 0;
            
            document.getElementById('first-price').textContent = formatCurrency(firstPrice);
            document.getElementById('last-price').textContent = formatCurrency(lastPrice);
            const changeEl = document.getElementById('price-change');
            changeEl.textContent = `${priceChangePct >= 0 ? '+' : ''}${priceChangePct.toFixed(2)}%`;
            changeEl.className = `text-xl font-semibold ${priceChangePct >= 0 ? 'text-green-600' : 'text-red-600'}`;
        }
    } catch (error) {
        console.error('Error loading price history metrics:', error);
    }
}

// Render research articles
function renderResearchArticles(articles) {
    if (!articles || articles.length === 0) {
        return;
    }
    
    const section = document.getElementById('research-section');
    section.classList.remove('section-hidden');
    
    document.getElementById('research-count').textContent = `Found ${articles.length} articles mentioning ${currentTicker} (last 30 days)`;
    
    const list = document.getElementById('research-articles-list');
    list.innerHTML = '';
    
    articles.slice(0, 10).forEach(article => {
        const articleDiv = document.createElement('div');
        articleDiv.className = 'border-b border-gray-200 py-4';
        
        const title = article.title || 'Untitled';
        const summary = article.summary || '';
        const url = article.url || '#';
        const source = article.source || 'Unknown';
        const publishedAt = formatDate(article.published_at);
        const sentiment = article.sentiment || 'N/A';
        
        const summaryId = `summary-${article.id || Math.random().toString(36).substr(2, 9)}`;
        const isLongSummary = summary.length > 500;
        const shortSummary = isLongSummary ? summary.substring(0, 500) + '...' : summary;
        
        articleDiv.innerHTML = `
            <details class="cursor-pointer">
                <summary class="font-semibold text-blue-600 hover:text-blue-800">${title}</summary>
                <div class="mt-2 pl-4">
                    <div id="${summaryId}-short" class="text-gray-700 mb-2">${shortSummary}</div>
                    ${isLongSummary ? `
                        <div id="${summaryId}-full" class="hidden text-gray-700 mb-2 whitespace-pre-wrap">${summary}</div>
                        <button onclick="toggleSummary('${summaryId}')" class="text-blue-600 hover:text-blue-800 text-sm font-medium mb-2">
                            <span id="${summaryId}-toggle">Show Full Summary</span>
                        </button>
                    ` : ''}
                    <div class="flex justify-between items-center text-sm text-gray-500">
                        <div>
                            <span>Source: ${source}</span>
                            ${publishedAt ? `<span class="ml-4">Published: ${publishedAt}</span>` : ''}
                            ${sentiment !== 'N/A' ? `<span class="ml-4">Sentiment: ${sentiment}</span>` : ''}
                        </div>
                        <a href="${url}" target="_blank" rel="noopener noreferrer" class="text-blue-600 hover:text-blue-800">Read Full Article →</a>
                    </div>
                </div>
            </details>
        `;
        list.appendChild(articleDiv);
    });
}

// Render social sentiment
function renderSocialSentiment(sentiment) {
    if (!sentiment) {
        return;
    }
    
    const section = document.getElementById('sentiment-section');
    section.classList.remove('section-hidden');
    
    // Render metrics
    if (sentiment.latest_metrics && sentiment.latest_metrics.length > 0) {
        const tbody = document.getElementById('sentiment-tbody');
        tbody.innerHTML = '';
        
        sentiment.latest_metrics.forEach(metric => {
            const row = document.createElement('tr');
            row.innerHTML = `
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${(metric.platform || 'N/A').charAt(0).toUpperCase() + (metric.platform || '').slice(1)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${metric.sentiment_label || 'N/A'}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${(metric.sentiment_score || 0).toFixed(2)}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${metric.volume || 0}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${metric.bull_bear_ratio !== null && metric.bull_bear_ratio !== undefined ? metric.bull_bear_ratio.toFixed(2) : 'N/A'}</td>
                <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-500">${formatDate(metric.created_at)}</td>
            `;
            tbody.appendChild(row);
        });
        
        document.getElementById('sentiment-metrics-container').style.display = 'block';
    } else {
        document.getElementById('sentiment-metrics-container').style.display = 'none';
    }
    
    // Render alerts
    if (sentiment.alerts && sentiment.alerts.length > 0) {
        const alertsList = document.getElementById('sentiment-alerts-list');
        alertsList.innerHTML = '';
        
        sentiment.alerts.forEach(alert => {
            const alertDiv = document.createElement('div');
            const platform = (alert.platform || 'Unknown').charAt(0).toUpperCase() + (alert.platform || '').slice(1);
            const sentimentLabel = alert.sentiment_label || 'N/A';
            const score = (alert.sentiment_score || 0).toFixed(2);
            
            let alertClass = 'bg-blue-100 border-blue-400 text-blue-700';
            if (sentimentLabel === 'EUPHORIC') {
                alertClass = 'bg-green-100 border-green-400 text-green-700';
            } else if (sentimentLabel === 'FEARFUL') {
                alertClass = 'bg-red-100 border-red-400 text-red-700';
            } else if (sentimentLabel === 'BULLISH') {
                alertClass = 'bg-blue-100 border-blue-400 text-blue-700';
            }
            
            alertDiv.className = `border px-4 py-3 rounded mb-2 ${alertClass}`;
            alertDiv.textContent = `${platform} - ${sentimentLabel} (Score: ${score})`;
            alertsList.appendChild(alertDiv);
        });
        
        document.getElementById('sentiment-alerts-container').style.display = 'block';
    } else {
        document.getElementById('sentiment-alerts-container').style.display = 'none';
    }
}

// Render congress trades
function renderCongressTrades(trades) {
    if (!trades || trades.length === 0) {
        return;
    }
    
    const section = document.getElementById('congress-section');
    section.classList.remove('section-hidden');
    
    document.getElementById('congress-count').textContent = `Found ${trades.length} recent trades by politicians (last 30 days)`;
    
    const tbody = document.getElementById('congress-tbody');
    tbody.innerHTML = '';
    
    trades.slice(0, 20).forEach(trade => {
        const row = document.createElement('tr');
        row.innerHTML = `
            <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${formatDate(trade.transaction_date)}</td>
            <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.politician || 'N/A'}</td>
            <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.chamber || 'N/A'}</td>
            <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.type || 'N/A'}</td>
            <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.amount || 'N/A'}</td>
            <td class="px-6 py-4 whitespace-nowrap text-sm text-gray-900">${trade.party || 'N/A'}</td>
        `;
        tbody.appendChild(row);
    });
}

// Render watchlist status
function renderWatchlistStatus(status) {
    if (!status) {
        return;
    }
    
    const section = document.getElementById('watchlist-section');
    section.classList.remove('section-hidden');
    
    document.getElementById('watchlist-status').textContent = status.is_active ? '✅ In Watchlist' : '❌ Not Active';
    document.getElementById('watchlist-tier').textContent = status.priority_tier || 'N/A';
    document.getElementById('watchlist-source').textContent = status.source || 'N/A';
}

// Utility functions
function formatDate(dateStr) {
    if (!dateStr) return 'N/A';
    try {
        const date = new Date(dateStr);
        return date.toLocaleDateString();
    } catch (e) {
        return dateStr.substring(0, 10); // Return first 10 chars if parsing fails
    }
}

function formatCurrency(value) {
    return `$${parseFloat(value || 0).toFixed(2)}`;
}

function formatNumber(value, decimals = 2) {
    return parseFloat(value || 0).toFixed(decimals);
}

function showLoading() {
    document.getElementById('loading-spinner').classList.remove('section-hidden');
}

function hideLoading() {
    document.getElementById('loading-spinner').classList.add('section-hidden');
}

function showError(message) {
    document.getElementById('error-text').textContent = message;
    document.getElementById('error-message').classList.remove('section-hidden');
}

function hideError() {
    document.getElementById('error-message').classList.add('section-hidden');
}

function toggleSummary(summaryId) {
    const shortDiv = document.getElementById(`${summaryId}-short`);
    const fullDiv = document.getElementById(`${summaryId}-full`);
    const toggleBtn = document.getElementById(`${summaryId}-toggle`);
    
    if (shortDiv && fullDiv && toggleBtn) {
        if (fullDiv.classList.contains('hidden')) {
            // Show full summary
            shortDiv.classList.add('hidden');
            fullDiv.classList.remove('hidden');
            toggleBtn.textContent = 'Show Less';
        } else {
            // Show short summary
            shortDiv.classList.remove('hidden');
            fullDiv.classList.add('hidden');
            toggleBtn.textContent = 'Show Full Summary';
        }
    }
}

function showPlaceholder() {
    document.getElementById('placeholder-message').classList.remove('section-hidden');
    hideAllSections();
}

function hidePlaceholder() {
    document.getElementById('placeholder-message').classList.add('section-hidden');
}

function hideAllSections() {
    const sections = [
        'basic-info-section',
        'external-links-section',
        'portfolio-section',
        'chart-section',
        'research-section',
        'sentiment-section',
        'congress-section',
        'watchlist-section'
    ];
    
    sections.forEach(id => {
        document.getElementById(id).classList.add('section-hidden');
    });
}
